components/rmrk/shoppingActions.ts

```diff
@@ -0,0 +1,159 @@
+import Connector from '@kodadot1/sub-api'
+import exec, { execResultValue, txCb } from '@/utils/transactionExecutor'
+import { notificationTypes, showNotification } from '@/utils/notification'
+import { unpin } from '@/utils/proxy'
+import { somePercentFromTX } from '@/utils/support'
+import nftById from '@/queries/nftById.graphql'
+
+export const enum ShoppingAction {
+  SEND = 'SEND',
+  CONSUME = 'CONSUME',
+  LIST = 'LIST',
+  BUY = 'BUY',
+}
+export const ownerActions = [
+  ShoppingAction.SEND,
+  ShoppingAction.CONSUME,
+  ShoppingAction.LIST,
+]
+export const buyActions = [ShoppingAction.BUY]
+
+export const getActions = function (
+  isOwner: boolean,
+  isAvailableToBuy: boolean
+) {
+  return isOwner ? ownerActions : isAvailableToBuy ? buyActions : []
+}
+
+const constructRmrk = (
+  selectedAction,
+  version,
+  meta,
+  nftId,
+  metaValid
+): string => {
+  return `RMRK::${selectedAction}::${version}::${nftId}${
+    metaValid ? '::' + meta : ''
+  }`
+}
+
+export const checkBuyBeforeSubmit = async ({
+  urlPrefix,
+  nftId,
+  currentOwnerId,
+  price,
+  action,
+  apollo,
+}) => {
+  const nft = await apollo.query({
+    query: nftById,
+    client: urlPrefix,
+    variables: {
+      id: nftId,
+    },
+  })
+
+  const {
+    data: { nFTEntity },
+  } = nft
+
+  if (
+    !nFTEntity ||
+    nFTEntity.currentOwner !== currentOwnerId ||
+    nFTEntity.burned ||
+    nFTEntity.price === 0 ||
+    nFTEntity.price !== price
+  ) {
+    showNotification(
+      `[RMRK::${action}] Owner changed or NFT does not exist`,
+      notificationTypes.warn
+    )
+    throw new ReferenceError('NFT has changed')
+  }
+}
+
+export const submitAction = async ({
```

Comment:
`metaMixin` is your friend

Reply:
Haven't seen such a function before : ). I will change it.